spec_integration/queue_runner.rb

```diff
@@ -0,0 +1,75 @@
+require 'knapsack_pro'
+require 'json'
+
+ENV['KNAPSACK_PRO_CI_NODE_BUILD_ID'] = SecureRandom.uuid
+ENV['KNAPSACK_PRO_TEST_SUITE_TOKEN_RSPEC'] = 'fake-api-token'
+ENV['KNAPSACK_PRO_TEST_DIR'] = 'spec_integration'
+ENV['KNAPSACK_PRO_TEST_FILE_PATTERN'] = "spec_integration/**{,/*/**}/*_spec.rb"
+
+RSPEC_OPTIONS = ENV.fetch('TEST__RSPEC_OPTIONS')
+SHOW_DEBUG_LOG = ENV['TEST__SHOW_DEBUG_LOG'] == 'true'
```

Comment:
Yes, I use it all the time when I create or edit the integration tests. I use debug log level so I can see the output from the RSpec test suite that I'm testing as part of the test example in `spec/integration/runners/queue/rspec_runner_spec.rb`.

When I'm done I disable debug log level. So that when you run RSpec tests you see clean documentation without noise from each RSpec run that's hidden for each test example.

Clean (debug log level off):

```
➜  knapsack_pro-ruby git:(refactor-queue-rspec-runner) bundle exec rspec spec/integration/runners/queue/rspec_runner_spec.rb

Randomized with seed 21262

KnapsackPro::Runners::Queue::RSpecRunner - Integration tests
  when a few batches of tests returned by the Queue API
    runs tests
    detects test execution times correctly before sending it to API
  when the RSpec seed is used
    is expected to eq 0
  when the test suite has pending tests
    shows the summary of pending tests
  when the RSpec split by examples is enabled AND --tag is set
    sets 1 as exit code AND raises an error (a test example path as a_spec.rb[1:1] would always be executed even when it does not have the tag that is set via the --tag option. We cannot run tests because it could lead to running unintentional tests)
  when filter_run_when_matching is set to :focus and some tests are tagged with the focus tag
    shows an error message for :focus tagged tests AND sets 1 as exit code (shows the error because the batch of tests that has no focus tagged tests will run tests instead of not running them)
  when the RSpec split by examples is enabled AND simplecov is used
    produces a code coverage report
  when RSpec options are not set
    uses a default progress formatter
  when the example_status_persistence_file_path option is used and multiple batches of tests are fetched from the Queue API and some tests are pending and failing
    runs tests AND creates the example status persistence file
  when a spec file has a syntax error outside of the test example
    stops running tests on the batch that has a test file with the syntax error AND returns 1 as exit code
  when --tag is set
    runs only tagged test examples from multiple batches of tests fetched from the Queue API
  when a test file raises an exception that cannot be handle by RSpec AND --error-exit-code is set
    sets a custom exit code
  when an invalid RSpec option is set
    returns 1 as exit code AND shows an error message to stderr
  when the fail_if_no_examples option is true AND a batch of tests has a test file without test examples
    sets 0 as exit code to ignore the fail_if_no_examples option
  when a termination signal is received by the process AND --error-exit-code is set
    terminates the process AND sets a custom exit code
  when the RSpec split by examples is enabled
    splits slow test files by examples AND ensures the test examples are executed only once
  when a failing test in a batch of tests that is not the last batch fetched from the Queue API
    returns 1 as exit code (it remembers that one of the batches has a failing test)
  when the RSpec split by examples is enabled AND JUnit XML formatter is used
    produces a JUnit XML report
  when --profile is set
    shows top slowest examples AND top slowest example groups
  when --fail-fast is set
    stops running tests on the failing test AND returns 1 as exit code AND shows a warning message
  when tests are failing AND --failure-exit-code is set
    returns a custom exit code
  when a test file raises an exception that cannot be handle by RSpec
    stops running tests when unhandled exception happens AND sets 1 as exit code AND shows summary of unexecuted tests
  when the fail_fast option is set with a specific number of tests
    stops running tests on the 2nd failing test AND returns 1 as exit code AND shows a warning message when fail fast limit met
  when the fail_if_no_examples option is true AND the late CI node has an empty batch of tests because other CI nodes already consumed tests from the Queue API
    sets 0 as exit code to ignore the fail_if_no_examples option
  when the late CI node has an empty batch of tests because other CI nodes already consumed tests from the Queue API
    sets 0 as exit code
  when the RSpec split by examples is enabled AND JSON formatter is used
    produces a JSON report
  when a syntax error (an exception) in spec_helper.rb
    exits early with 1 as the exit code without running tests because RSpec wants to quit
  when deprecated run_all_when_everything_filtered option is true
    shows an error message AND sets 1 as exit code
  when hooks are defined
    calls RSpec before/after hooks only once for multiple batches of tests
    calls queue hooks for multiple batches of tests (queue hooks can be defined multiple times)
    calls hooks defined with when_first_matching_example_defined only once for multiple batches of tests
  when a failing test raises an exception
    returns 1 as exit code AND the exception does not leak outside of the RSpec runner context
  when RSpec is quitting
    returns non zero exit code because RSpec is quitting
  when RSpec options are not set AND Knapsack Pro log level is warn
    uses a default progress formatter AND shows dots for all test examples
  when a termination signal is received by the process
    terminates the process after tests from the current RSpec ExampleGroup are executed and sets 1 as exit code
  when spec_helper.rb has a missing KnapsackPro::Adapters::RSpecAdapter.bind method
    is expected to eq 1

Finished in 8.25 seconds (files took 0.4185 seconds to load)
36 examples, 0 failures

Randomized with seed 21262
```

With noise (debug log level ON):

```
➜  knapsack_pro-ruby git:(refactor-queue-rspec-runner) bundle exec rspec spec/integration/runners/queue/rspec_runner_spec.rb

Randomized with seed 55843

KnapsackPro::Runners::Queue::RSpecRunner - Integration tests
  when the fail_if_no_examples option is true AND a batch of tests has a test file without test examples
==================================================
STDOUT:
D, [2024-02-09T12:24:50.965696 #1895] DEBUG -- : [knapsack_pro] Test suite time execution queue recording enabled.
DEBUG: mocked tests from the Queue API: ["spec_integration/a_spec.rb"]

A_describe
  A1 test example
I, [2024-02-09T12:24:50.980568 #1895]  INFO -- : [knapsack_pro] To retry the last batch of tests fetched from the Queue API, please run the following command on your machine:
I, [2024-02-09T12:24:50.980585 #1895]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --default-path spec_integration "spec_integration/a_spec.rb"
DEBUG: mocked tests from the Queue API: ["spec_integration/b_spec.rb"]
I, [2024-02-09T12:24:50.981279 #1895]  INFO -- : [knapsack_pro] To retry the last batch of tests fetched from the Queue API, please run the following command on your machine:
I, [2024-02-09T12:24:50.981291 #1895]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --default-path spec_integration "spec_integration/b_spec.rb"
DEBUG: mocked tests from the Queue API: ["spec_integration/c_spec.rb"]

C_describe
  C1 test example
I, [2024-02-09T12:24:50.982606 #1895]  INFO -- : [knapsack_pro] To retry the last batch of tests fetched from the Queue API, please run the following command on your machine:
I, [2024-02-09T12:24:50.982616 #1895]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --default-path spec_integration "spec_integration/c_spec.rb"
DEBUG: mocked tests from the Queue API: []
D, [2024-02-09T12:24:50.982692 #1895] DEBUG -- : [knapsack_pro] Global time execution for tests: 0.016231999994488433s

Finished in 0.01623 seconds (files took 0.02661 seconds to load)
2 examples, 0 failures

I, [2024-02-09T12:24:50.983368 #1895]  INFO -- : [knapsack_pro] To retry all the tests assigned to this CI node, please run the following command on your machine:
I, [2024-02-09T12:24:50.983376 #1895]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --default-path spec_integration "spec_integration/a_spec.rb" "spec_integration/b_spec.rb" "spec_integration/c_spec.rb"
DEBUG: mocked the create_build_subset method

==================================================
STDERR:


==================================================
Exit status code:
pid 1895 exit 0

    sets 0 as exit code to ignore the fail_if_no_examples option
  when spec_helper.rb has a missing KnapsackPro::Adapters::RSpecAdapter.bind method
==================================================
STDOUT:
DEBUG: mocked tests from the Queue API: ["spec_integration/a_spec.rb"]
.I, [2024-02-09T12:24:51.206410 #1898]  INFO -- : [knapsack_pro] To retry the last batch of tests fetched from the Queue API, please run the following command on your machine:
I, [2024-02-09T12:24:51.206446 #1898]  INFO -- : [knapsack_pro] bundle exec rspec --format progress --default-path spec_integration "spec_integration/a_spec.rb"
DEBUG: mocked tests from the Queue API: ["spec_integration/b_spec.rb"]
.I, [2024-02-09T12:24:51.207272 #1898]  INFO -- : [knapsack_pro] To retry the last batch of tests fetched from the Queue API, please run the following command on your machine:
I, [2024-02-09T12:24:51.207281 #1898]  INFO -- : [knapsack_pro] bundle exec rspec --format progress --default-path spec_integration "spec_integration/b_spec.rb"
DEBUG: mocked tests from the Queue API: []


Finished in 0.01617 seconds (files took 0.0258 seconds to load)
2 examples, 0 failures

I, [2024-02-09T12:24:51.207894 #1898]  INFO -- : [knapsack_pro] To retry all the tests assigned to this CI node, please run the following command on your machine:
I, [2024-02-09T12:24:51.207903 #1898]  INFO -- : [knapsack_pro] bundle exec rspec --format progress --default-path spec_integration "spec_integration/a_spec.rb" "spec_integration/b_spec.rb"
DEBUG: mocked the create_build_subset method


E, [2024-02-09T12:24:51.207983 #1898] ERROR -- : [knapsack_pro] ---------- Configuration error --------------------------------------------------
E, [2024-02-09T12:24:51.207990 #1898] ERROR -- : [knapsack_pro] You forgot to call KnapsackPro::Adapters::RSpecAdapter.bind method in your test runner configuration file. It is needed to record test files time execution. Please follow the installation guide to configure your project properly https://knapsackpro.com/perma/ruby/installation-guide
E, [2024-02-09T12:24:51.207995 #1898] ERROR -- : [knapsack_pro] If you already have KnapsackPro::Adapters::RSpecAdapter.bind method added and you still see this error then one of your tests must have deleted the .knapsack_pro directory from the disk accidentally. Please ensure you do not remove the .knapsack_pro directory: https://knapsackpro.com/perma/ruby/dashboard-zeroish-test-execution-times

==================================================
STDERR:


==================================================
Exit status code:
pid 1898 exit 1

    is expected to eq 1
  when the fail_if_no_examples option is true AND the late CI node has an empty batch of tests because other CI nodes already consumed tests from the Queue API
==================================================
STDOUT:
D, [2024-02-09T12:24:51.427902 #1900] DEBUG -- : [knapsack_pro] Test suite time execution queue recording enabled.
DEBUG: mocked tests from the Queue API: []
D, [2024-02-09T12:24:51.430190 #1900] DEBUG -- : [knapsack_pro] Global time execution for tests: 0.0006460000004153699s

Finished in 0.00066 seconds (files took 0.03892 seconds to load)
0 examples, 0 failures

W, [2024-02-09T12:24:51.431219 #1900]  WARN -- : [knapsack_pro] No test files were executed on this CI node.
D, [2024-02-09T12:24:51.431235 #1900] DEBUG -- : [knapsack_pro] This CI node likely started work late after the test files were already executed by other CI nodes consuming the queue.
DEBUG: mocked the create_build_subset method

==================================================
STDERR:


==================================================
Exit status code:
pid 1900 exit 0

    sets 0 as exit code to ignore the fail_if_no_examples option
  when deprecated run_all_when_everything_filtered option is true
==================================================
STDOUT:
D, [2024-02-09T12:24:51.667592 #1903] DEBUG -- : [knapsack_pro] Test suite time execution queue recording enabled.
E, [2024-02-09T12:24:51.668398 #1903] ERROR -- : [knapsack_pro] The run_all_when_everything_filtered option is deprecated. See: https://knapsackpro.com/perma/ruby/rspec-deprecated-run-all-when-everything-filtered

==================================================
STDERR:
/Users/artur/Documents/github/knapsack-pro/knapsack_pro-ruby/lib/knapsack_pro/runners/queue/rspec_runner.rb:37:in `ensure_no_deprecated_run_all_when_everything_filtered_option!': The run_all_when_everything_filtered option is deprecated. See: https://knapsackpro.com/perma/ruby/rspec-deprecated-run-all-when-everything-filtered (RuntimeError)
	from /Users/artur/Documents/github/knapsack-pro/knapsack_pro-ruby/lib/knapsack_pro/runners/queue/rspec_runner.rb:238:in `pre_run_setup'
	from /Users/artur/Documents/github/knapsack-pro/knapsack_pro-ruby/lib/knapsack_pro/runners/queue/rspec_runner.rb:170:in `run'
	from /Users/artur/Documents/github/knapsack-pro/knapsack_pro-ruby/lib/knapsack_pro/runners/queue/rspec_runner.rb:148:in `run'
	from spec_integration/queue_runner.rb:75:in `<main>'

==================================================
Exit status code:
pid 1903 exit 1

    shows an error message AND sets 1 as exit code
  when a failing test in a batch of tests that is not the last batch fetched from the Queue API
==================================================
STDOUT:
D, [2024-02-09T12:24:51.882015 #1906] DEBUG -- : [knapsack_pro] Test suite time execution queue recording enabled.
DEBUG: mocked tests from the Queue API: ["spec_integration/a_spec.rb", "spec_integration/failing_spec.rb"]

A_describe
  A1 test example

B_describe
  B1 test example (FAILED - 1)
I, [2024-02-09T12:24:51.903638 #1906]  INFO -- : [knapsack_pro] To retry the last batch of tests fetched from the Queue API, please run the following command on your machine:
I, [2024-02-09T12:24:51.903663 #1906]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --default-path spec_integration "spec_integration/a_spec.rb" "spec_integration/failing_spec.rb"
DEBUG: mocked tests from the Queue API: ["spec_integration/c_spec.rb"]

C_describe
  C1 test example
I, [2024-02-09T12:24:51.904606 #1906]  INFO -- : [knapsack_pro] To retry the last batch of tests fetched from the Queue API, please run the following command on your machine:
I, [2024-02-09T12:24:51.904619 #1906]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --default-path spec_integration "spec_integration/c_spec.rb"
DEBUG: mocked tests from the Queue API: []
D, [2024-02-09T12:24:51.904697 #1906] DEBUG -- : [knapsack_pro] Global time execution for tests: 0.021911999996518716s

Failures:

  1) B_describe B1 test example
     Failure/Error: expect(1).to eq 0

       expected: 0
            got: 1

       (compared using ==)
     # ./spec_integration/failing_spec.rb:3:in `block (2 levels) in <top (required)>'
     # ./lib/knapsack_pro/adapters/rspec_adapter.rb:103:in `block (2 levels) in ensure_no_focus!'
     # ./lib/knapsack_pro/extensions/rspec_extension.rb:124:in `block (4 levels) in knapsack__run_specs'
     # ./lib/knapsack_pro/extensions/rspec_extension.rb:122:in `map'
     # ./lib/knapsack_pro/extensions/rspec_extension.rb:122:in `block (3 levels) in knapsack__run_specs'
     # ./lib/knapsack_pro/runners/queue/rspec_runner.rb:204:in `block (2 levels) in with_batched_tests_from_queue'
     # ./lib/knapsack_pro/runners/queue/rspec_runner.rb:268:in `with_queue_hooks'
     # ./lib/knapsack_pro/runners/queue/rspec_runner.rb:203:in `block in with_batched_tests_from_queue'
     # ./lib/knapsack_pro/runners/queue/rspec_runner.rb:196:in `loop'
     # ./lib/knapsack_pro/runners/queue/rspec_runner.rb:196:in `with_batched_tests_from_queue'
     # ./lib/knapsack_pro/extensions/rspec_extension.rb:119:in `block (2 levels) in knapsack__run_specs'
     # ./lib/knapsack_pro/extensions/rspec_extension.rb:118:in `block in knapsack__run_specs'
     # ./lib/knapsack_pro/extensions/rspec_extension.rb:117:in `knapsack__run_specs'
     # ./lib/knapsack_pro/runners/queue/rspec_runner.rb:179:in `run'
     # ./lib/knapsack_pro/runners/queue/rspec_runner.rb:148:in `run'
     # spec_integration/queue_runner.rb:75:in `<main>'

Finished in 0.02192 seconds (files took 0.02655 seconds to load)
3 examples, 1 failure

Failed examples:

rspec ./spec_integration/failing_spec.rb:2 # B_describe B1 test example

I, [2024-02-09T12:24:51.917052 #1906]  INFO -- : [knapsack_pro] To retry all the tests assigned to this CI node, please run the following command on your machine:
I, [2024-02-09T12:24:51.917071 #1906]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --default-path spec_integration "spec_integration/a_spec.rb" "spec_integration/failing_spec.rb" "spec_integration/c_spec.rb"
DEBUG: mocked the create_build_subset method

==================================================
STDERR:


==================================================
Exit status code:
pid 1906 exit 1

    returns 1 as exit code (it remembers that one of the batches has a failing test)
  when the RSpec split by examples is enabled AND JUnit XML formatter is used
==================================================
STDOUT:
D, [2024-02-09T12:24:52.124043 #1910] DEBUG -- : [knapsack_pro] Detected 1 slow test files: [{"path"=>"spec_integration/a_spec.rb"}]
DEBUG: mocked test file cases for slow test files: [{"path"=>"spec_integration/a_spec.rb"}]
D, [2024-02-09T12:24:52.128032 #1910] DEBUG -- : [knapsack_pro] Test suite time execution queue recording enabled.
DEBUG: mocked tests from the Queue API: ["spec_integration/a_spec.rb[1:1]", "spec_integration/b_spec.rb"]

A_describe
  A1 test example

B_describe
  B1 test example
  B2 test example
I, [2024-02-09T12:24:52.147472 #1910]  INFO -- : [knapsack_pro] To retry the last batch of tests fetched from the Queue API, please run the following command on your machine:
I, [2024-02-09T12:24:52.147501 #1910]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --format RspecJunitFormatter --out ./spec_integration/rspec.xml --default-path spec_integration "spec_integration/a_spec.rb[1:1]" "spec_integration/b_spec.rb"
DEBUG: mocked tests from the Queue API: ["spec_integration/a_spec.rb[1:2]", "spec_integration/c_spec.rb"]

A_describe
  A2 test example

C_describe
  C1 test example
  C2 test example
I, [2024-02-09T12:24:52.149092 #1910]  INFO -- : [knapsack_pro] To retry the last batch of tests fetched from the Queue API, please run the following command on your machine:
I, [2024-02-09T12:24:52.149102 #1910]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --format RspecJunitFormatter --out ./spec_integration/rspec.xml --default-path spec_integration "spec_integration/a_spec.rb[1:2]" "spec_integration/c_spec.rb"
DEBUG: mocked tests from the Queue API: []
D, [2024-02-09T12:24:52.149181 #1910] DEBUG -- : [knapsack_pro] Global time execution for tests: 0.017801999987568706s

Finished in 0.01781 seconds (files took 0.03064 seconds to load)
6 examples, 0 failures

I, [2024-02-09T12:24:52.150182 #1910]  INFO -- : [knapsack_pro] To retry all the tests assigned to this CI node, please run the following command on your machine:
I, [2024-02-09T12:24:52.150195 #1910]  INFO -- : [knapsack_pro] bundle exec rspec --format documentation --format RspecJunitFormatter --out ./spec_integration/rspec.xml --default-path spec_integration "spec_integration/a_spec.rb[1:1]" "spec_integration/b_spec.rb" "spec_integration/a_spec.rb[1:2]" "spec_integration/c_spec.rb"
DEBUG: mocked the create_build_subset method

==================================================
STDERR:


==================================================
Exit status code:
pid 1910 exit 0

    produces a JUnit XML report

.....
```